streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -2693,137 +2693,110 @@ <VT, VR> KStream<K, VR> leftJoin(final KTable<K, VT> table,
                                      final Joined<K, V, VT> joined);
 
     /**
-     * Join records of this stream with {@link GlobalKTable}'s records using non-windowed inner equi join.
+     * Join records of this stream with {@link GlobalKTable}'s records using non-windowed inner equi-join.
      * The join is a primary key table lookup join with join attribute
-     * {@code keyValueMapper.map(stream.keyValue) == table.key}.
+     * {@code keyValueMapper.map(streamRecord) == tableRecord.key}.
      * "Table lookup join" means, that results are only computed if {@code KStream} records are processed.
-     * This is done by performing a lookup for matching records in the <em>current</em> internal {@link GlobalKTable}
-     * state.
+     * This is done by performing a lookup for matching records in the <em>current</em> (i.e., processing time)
+     * internal {@link GlobalKTable} state.
      * In contrast, processing {@link GlobalKTable} input records will only update the internal {@link GlobalKTable}
      * state and will not produce any result records.
-     * <p>
-     * For each {@code KStream} record that finds a corresponding record in {@link GlobalKTable} the provided
+     *
+     * <p>For each {@code KStream} record that finds a joining record in the {@link GlobalKTable} the provided
      * {@link ValueJoiner} will be called to compute a value (with arbitrary type) for the result record.
-     * The key of the result record is the same as the key of this {@code KStream}.
-     * If a {@code KStream} input value is {@code null} the record will not be included in the join operation
-     * and thus no output record will be added to the resulting {@code KStream}.
+     * The key of the result record is the same as the stream record's key.
+     * If you need read access to the {@code KStream} key, use {@link #join(GlobalKTable, KeyValueMapper, ValueJoinerWithKey)}.
```

Comment:
I just describe what the code currently does. But it seems it's not correct. We should not pass in the KStream record key, but the join key (ie, whatever `keySelector` returns).

The original KIP says (https://cwiki.apache.org/confluence/display/KAFKA/KIP-149%3A+Enabling+key+access+in+ValueTransformer%2C+ValueMapper%2C+and+ValueJoiner)
> it seems like extending the interface to pass the join key along as well would be helpful

The question now is, can we change this directly? Even if it would be a bug-fix, is could be considered a "breaking change" (should we try to sneak a fix into `4.0` which would allows us to ship a breaking change)?